src/misc/unicode.cpp

```diff
@@ -0,0 +1,1129 @@
+                                     CodePageMappingReverse &mapping)
+{
+    // Import code page character -> UTF-8 mapping from external file
+
+    // Open the file
+    auto in_file = open_mapping_file(path_root, file_name);
+    if (!in_file) {
+        LOG_ERR("UTF8: Error opening mapping file %s", file_name.c_str());
+        return false;
+    }
+
+    // Read and parse
+    std::string line_str = " ";
+    size_t      line_num = 0;
+    std::smatch match;
+
+    CodePageMappingReverse new_mapping;
+
+    while (get_mapping_line(in_file, line_str, line_num)) {
+
+        if (line_str.length() == 1 && line_str[0] == 0x1a)
+            break; // end of file marking, for some fiels from unicode.org
+
+        auto get_hex_match = [&match](const uint8_t i) {
+            return std::strtol(match.str(i).c_str(), nullptr, 16);
+        };
+
+        if (std::regex_match(line_str, match, regex_map_code) ||
+            std::regex_match(line_str, match, regex_map_code_mark1) ||
+            std::regex_match(line_str, match, regex_map_code_mark2)) {
+
+            // Handle mapping entry
+
+            const auto character_code = static_cast<uint8_t>(get_hex_match(1));
+
+            if (character_code >= 0x80) { // ignore codes below 0x80
+                const auto code_point = static_cast<uint16_t>(get_hex_match(2));
+
+                Grapheme grapheme(code_point);
+                if (match.size() > 3)
+                    grapheme.AddMark(static_cast<uint16_t>(get_hex_match(3)));
+                if (match.size() > 4)
+                    grapheme.AddMark(static_cast<uint16_t>(get_hex_match(4)));
+
+                // Invalid grapheme that is not added (overridden) is OK here;
+                // at least CP 1258 definition from Unicode.org contains mapping
+                // of code page characters to combining marks, which is fine
+                // for converting texts, but a no-no for DOS emulation (where
+                // the number of output characters has to match the number of
+                // input characters).
+                // For such code page definitions, just override problematic
+                // mappings in the main mapping configuration file.
+
+                if (add_if_not_mapped(new_mapping, character_code, grapheme) &&
+                    !check_grapheme_valid(grapheme, file_name, line_num))
+                    return false;
+            }
+
+        } else if (std::regex_match(line_str, match, regex_map_undefined)) {
+
+            // Handle undefined character entry
+
+            const auto character_code = static_cast<uint8_t>(get_hex_match(1));
+            if (character_code >= 0x80)  { // ignore codes below 0x80
+                Grapheme grapheme;
+                add_if_not_mapped(new_mapping, character_code, grapheme);
+            }
+
+        } else {
+            error_parsing(file_name, line_num);
+            return false;
+        }
```

Comment:
minor: could get rid of quite a few newlines in this function

Reply:
OK, reduced the amount of free lines,